### PR TITLE
pass featurizer kwargs

### PR DESCRIPTION
**Pull request by walid0925**
# Pull Request Template

## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/feat/base_classes.py [hunk 1]
```diff
@@ -23,10 +23,14 @@ class Featurizer(object):
   new datatype.
   """
 
-  def featurize(self, datapoints: Iterable[Any],
-                log_every_n: int = 1000) -> np.ndarray:
+  def featurize(self,
```
Comment (rbharath) [inline deepchem/feat/base_classes.py:26]: @walid0925 brought this up offline already so just documenting for the record. Mypy is complaining that we need to add `kwargs` on to the other subclasses so we should do that here before merging in

**Review by rbharath — COMMENTED** [review]
Should be good to go once we fix mypy as @walid0925 has already pointed out (see below0

CC @seyonechithrananda

**Comment by walid0925** [comment]
I'm having some issues getting `mypy` to cooperate. it seems like even with adding `**kwargs` to the subclass method, `mypy` still complains that the signature is incompatible. for example, changing 

```
class ComplexFeaturizer(Featurizer):
  """"
  Abstract class for calculating features for mol/protein complexes.
  """
  def featurize(self,
                complexes: Iterable[Tuple[str, str]],
                log_every_n: int = 100,
                **kwargs: Any) -> np.ndarray:
    """
...
```

still gives

```
deepchem/feat/base_classes.py:162: error: Signature of "featurize" incompatible with supertype "Featurizer"
```

would appreciate any input or suggestions!

**Comment by peastman** [comment]
Perhaps it's objecting that you have a different default value for `log_every_n`?

**Comment by walid0925** [comment]
> Perhaps it's objecting that you have a different default value for `log_every_n`?

After removing `**kwargs` the `mypy` error goes away, so it seems to be directly because of `**kwargs`

**Comment by walid0925** [comment]
hi @rbharath and @peastman - bumping this since I haven't found a good solution. I've also tried using `# type: ignore` but it hasn't seemed to help

cc @seyonechithrananda

**Comment by peastman** [comment]
Featurizer defines it as

```python
  def featurize(self,
                datapoints: Iterable[Any],
                log_every_n: int = 1000,
                **kwargs) -> np.ndarray:
```

ComplexFeaturizer defines it as

```python
  def featurize(self,
                complexes: Iterable[Tuple[str, str]],
                log_every_n: int = 100,
                **kwargs) -> np.ndarray:
```

I think it's objecting that the name of the first argument is different.

**Comment by walid0925** [comment]
> Featurizer defines it as
> 
> ```python
>   def featurize(self,
>                 datapoints: Iterable[Any],
>                 log_every_n: int = 1000,
>                 **kwargs) -> np.ndarray:
> ```
> 
> ComplexFeaturizer defines it as
> 
> ```python
>   def featurize(self,
>                 complexes: Iterable[Tuple[str, str]],
>                 log_every_n: int = 100,
>                 **kwargs) -> np.ndarray:
> ```
> 
> I think it's objecting that the name of the first argument is different.

@peastman the puzzling thing is that having

Featurizer
```python
  def featurize(self, datapoints: Iterable[Any],
                log_every_n: int = 1000) -> np.ndarray:
```

and ComplexFeaturizer
```python
  def featurize(self,
                complexes: Iterable[Tuple[str, str]],
                log_every_n: int = 100) -> np.ndarray:
```
raises no errors

**Comment by peastman** [comment]
I can only guess what logic it's applying.  I suspect it's something like this:

The first version includes `datapoints` and `kwargs`.  The second includes `complexes` and `kwargs`.  What happens if a user calls it specifying both `datapoints` and `kwargs` by name?  That's a valid thing to do in both cases, but they'll disagree on which should be passed as the first argument and which should be included in `kwargs`.

Of course, if you don't have `kwargs`, the signatures are still inconsistent because the first argument might still get specified by name.  But in that case, you'll immediately discover the error because it will throw an exception.  So maybe the Mypy authors decided not to flag that case as an error.

**Comment by walid0925** [comment]
> I can only guess what logic it's applying. I suspect it's something like this:
> 
> The first version includes `datapoints` and `kwargs`. The second includes `complexes` and `kwargs`. What happens if a user calls it specifying both `datapoints` and `kwargs` by name? That's a valid thing to do in both cases, but they'll disagree on which should be passed as the first argument and which should be included in `kwargs`.
> 
> Of course, if you don't have `kwargs`, the signatures are still inconsistent because the first argument might still get specified by name. But in that case, you'll immediately discover the error because it will throw an exception. So maybe the Mypy authors decided not to flag that case as an error.

thanks @peastman ! I think you hit the nail on the head here and it helped us resolve the issue by changing the the name of the `complexes` argument in #2581

**Pull request merged**